Moved accounts logic to PCAWrapper and added tests.

shalinikhare27 opened the pull request:
# What
Extended the IPCAWrapper to implement `TryToGetCachedAccountAsync` and `TryToGetCachedAccountsAsync` method.

# Why
We try to get a cached account to auth silently when possible. In the current design for `TokenFetcherPublicClient` it tries to fetch the account right before running `GetTokenNormalFlowAsync` and is specifically used by Broker and Web auth flow. Since going forward, we will also use it for DeviceCodeFlow - where we try to auth silently, I wanted to extract this method to be used more easily.

# Things to note
`TryToGetCachedAccountAsync` - Once it gets a list of cached accounts, this method determine if we should return `null` or an `Account`
`TryToGetCachedAccountsAsync` - This method calls the `GetAccountsAsync` method, part of the PCAWrapper class. It then determines the list of accounts found in the cache and then filters based on preferred domain if provided.
Creating two separate method was my preference as we might want to get a list of cached accounts in future. The current `TokenFetcherPublicClient` just has one method `TryToGetCachedAccountAsync` which essentially gets a list of accounts and then decides which account to pick. I just broke it into two more readable method. 

# Test
Nothing to test so far for integration testing. This is just a refactor.
Unit tests are added.

# Class diagram
For brevity, I have only represented `AuthFlowBroker` but you could have any auth flow and the flow will look the same more or less.

```mermaid
classDiagram
      IAuthFlows <|-- AuthFlows
      IAuthFlows <|-- AuthFlowBroker
      AuthFlowBroker*-- PCACache
      AuthFlowBroker*-- PCAWrapper
      AuthFlowBroker*-- TaskExecutor
     AuthFlowBroker*-- MsalHttpClientFactoryAdaptor
     AuthFlowBroker*-- TokenResultExtensions
      IAuthFlows : +Task<TokenResult> GetTokenAsync()
      PCAWrapper *-- IPublicClientApplication
      class AuthFlows{
          +ILogger logger
          +Guid resourceId
          +Guid clientId
          +Guid tenantId
          +IEnumerable<string> scopes = null
          +string osxKeyChainSuffix = null
          +string preferredDomain = null, 
          +bool verifyPersistence = false
      }
      class AuthFlowBroker{
          +ILogger logger
          +Guid clientId
          +Guid tenantId
          +IEnumerable<string> scopes = null
          +string osxKeyChainSuffix = null
          +string preferredDomain = null, 
          +bool verifyPersistence = false
      }
```

shalinikhare27 commented on src/MSALWrapper/MSALWrapper.csproj:
```diff
@@ -46,5 +46,10 @@
+        <_Parameter1>$(AssemblyName).Test</_Parameter1>
+        <!-- We use the value of AssemblyName to declare the value of the attribute -->
+    </AssemblyAttribute>
+  </ItemGroup>
```
removed with further refactoring.

shalinikhare27 commented on src/MSALWrapper/AccountsProvider/IAccountsProvider.cs:
```diff
@@ -0,0 +1,29 @@
+        /// </summary>
+        /// <param name="preferredDomain">The preferred domain.</param>
+        /// <returns>The <see cref="Task"/>.</returns>
+        Task<IList<IAccount>> TryGetAccountsAsync(string preferredDomain = null);
```
Please review the latest refactor. :)

shalinikhare27 commented:
> After discussing offline - I think we actually don't need this interface, and instead want to extend the existing IPCAWRrapper interface to also contain an abstraction over the GetAccounts methods there. The implementation on the PCAWrapper class would be the same, we just don't need/want to split up the abstraction over the MSAL PCA interface into multiple places.

Yep, implemented this in my latest changes. Please review again.

kyle-rader commented on src/MSALWrapper/PCAWrapper.cs:
```diff
@@ -83,6 +85,20 @@ public interface IPCAWrapper
+        Task<IAccount> TryToGetCachedAccountAsync(string preferredDomain = null);
+
+        /// <summary>
+        /// Tries to get a list of cached accounts using the preferred domain if provided.
```
Since this is a "Try*" method, typically that means it might fail. We want to document how this behaves if it fails as well.

shalinikhare27 replied:
Added.

kyle-rader commented on src/MSALWrapper/PCAWrapper.cs:
```diff
@@ -83,6 +85,20 @@ public interface IPCAWrapper
         Task<TokenResult> GetTokenDeviceCodeAsync(IEnumerable<string> scopes, Func<DeviceCodeResult, Task> callback, CancellationToken cancellationToken);
+
+        /// <summary>
+        /// Tries to get a cached account using the preferred domain if provided.
```
(Also what happens if this fails)

shalinikhare27 replied:
Gave it a shot. Please let me know if you want to write it differently.

kyle-rader commented on src/MSALWrapper/PCAWrapper.cs:
```diff
@@ -93,19 +109,29 @@ public interface IPCAWrapper
             this.pca = pca;
         }
 
+        /// <summary>
```
(I don't expect it to be that difficult to pass in a logger in the existing uses).

shalinikhare27 replied:
Are you suggesting the following:
`        public PCAWrapper(ILogger logger = null, IPublicClientApplication pca)
        {
            this.logger = logger;
            this.pca = pca;
        }
`

Just have on constructor like this and fix `TokenFetcherPublicClient` and other existing classes to take a logger in the `PCAWrapper` constrictor? Please elaborate so that I can fully understand how you want these. :)

kyle-rader replied:
Almost - just drop the `= null`. So 
```chsarp
public PCAWrapper(ILogger logger, IPublicClientApplication pca)
{
  this.logger = logger;
  this.pca = pca;
}
```

shalinikhare27 replied:
Done!

kyle-rader commented on src/MSALWrapper.Test/PCAWrapperTest.cs:
```diff
@@ -0,0 +1,210 @@
+            result.Should().BeNull();
+        }
+
+        private void GetAccountsMock(IEnumerable<IAccount> accounts)
```
Or my personal preference, to use verb-based naming `MockAccounts`

kyle-rader commented on src/MSALWrapper.Test/PCAWrapperTest.cs:
```diff
@@ -0,0 +1,210 @@
+            this.testAccount = new Mock<IAccount>(MockBehavior.Strict);
+            this.testAccount.Setup(a => a.Username).Returns(TestUser);
+
+            this.pcaWrapper = new Mock<IPCAWrapper>(MockBehavior.Strict);
```
Given this test class is testing the PCAWrapper class - we should not need any mocks of it ;)

kyle-rader reviewed (changes requested):
This is moving in the right direction! I've made a lot of small comments but overall I think this is a good extension of the `IPCAWrapper` interface.

reillysiemens commented on src/MSALWrapper/PCAWrapper.cs:
```diff
@@ -83,6 +85,22 @@ public interface IPCAWrapper
+
+        /// <summary>
+        /// Tries to return a cached account when the list returns only one account using the preferred domain if provided.
+        /// If there are more than one account returned in the list, then return null.
```
```suggestion
        /// If there is more than one account returned in the list, then return null.

```

reillysiemens replied:
Another way to phrase that might be

> A `null` return indicates more than one account was found.

Though, I'm not sure that's actually correct as it seems one of the tests here shows that `null` is also returned when the list of accounts is empty. Might be worth documenting that as well.

Also, it seems like even if there's more than one account, as long as there's only one matching the `preferredDomain` then that gets returned.

Everything else here looks good, but I think there's room to improve the documentation of this interface (and the corresponding implementation site.

shalinikhare27 replied:
Gave it a shot! Let me know what you think.

reillysiemens replied:
Looks good to me!

reillysiemens reviewed (changes requested):
This all looks great save for some minor docs improvements. Would be happy to slap a ✅ on this after that gets fixed up.

reillysiemens reviewed (approved):
With those comment fixes I think this is good to go. 🚀🚢

shalinikhare27 commented:
Hey @kyle-rader , I think I addressed all your comments and I have consensus from the rest of the team to move forward with this. I dismissed your review since you're busy, but feel free to reach out if I've left anything unaddressed here. I can create a follow up PR.
